index.ts

```diff
@@ -14,33 +14,28 @@ abstract class NgComponent<Props, State> {
     }
   */
   // nb: this method is explicity exposed for unit testing
-  public $onChanges(changes: {}) {
+  public $onChanges(changes: object) {
     const oldProps = mapValues<{}, Props>(changes, 'previousValue')
     const newProps = mapValues<{}, Props>(changes, 'currentValue')
 
-    if (!this.__isFirstRender) {
-      this.componentWillReceiveProps(newProps)
-    }
-
-    // store the new props
-    this.props = assign({}, this.props, newProps)
+    const props = assign({}, this.props, newProps)
+    const state = assign({}, this.state) // TODO: implement setState and test it with shouldComponentUpdate
```

Comment:
Is this to prevent reference equality between `state` and `this.state` in `shouldComponentUpdate`? If so, we should add a test.